runtime/memprof.c

```diff
@@ -156,1176 +1316,998 @@ static memprof_domain_t domain_create(caml_domain_state *caml_state)
-                Wosize_hp(p), SRC_MARSHAL, is_young, Val_hp(p), callstack);
-    p = next_p;
-  }
-  check_action_pending();
+  /* There should currently be just one thread in this domain */
+  CAMLassert(thread->next == NULL);
+  return thread;
 }
 
-/**** Interface with the OCaml code. ****/
+CAMLexport void caml_memprof_delete_thread(memprof_thread_t thread)
+{
+  /* Transfer entries to the domain. If this fails due to allocation
+   * failure, we will lose the entries. */
+  (void)entries_transfer(&thread->entries, &thread->domain->entries);
+  thread_destroy(thread);
+}
 
-static void caml_memprof_init(void)
+CAMLexport void caml_memprof_enter_thread(memprof_thread_t thread)
 {
-  init = 1;
-  xoshiro_init();
+  CAMLassert(thread);
+  thread->domain->current = thread;
+  update_suspended(thread->domain, thread->suspended);
 }
 
-CAMLprim value caml_memprof_start(value lv, value szv, value tracker_param)
+/**** Interface to OCaml ****/
+
+CAMLprim value caml_memprof_start(value lv, value szv, value tracker)
 {
-  CAMLparam3(lv, szv, tracker_param);
+  CAMLparam3(lv, szv, tracker);
+  CAMLlocal1(one_log1m_lambda_v);
 
-  double l = Double_val(lv);
+  double lambda = Double_val(lv);
   intnat sz = Long_val(szv);
 
-  if (started) caml_failwith("Gc.Memprof.start: already started.");
-
-  if (sz < 0 || !(l >= 0.) || l > 1.) /* Checks that [l] is not NAN. */
+  /* Checks that [lambda] is within range (and not NaN). */
+  if (sz < 0 || lambda < 0.0 || lambda > 1.0)
     caml_invalid_argument("Gc.Memprof.start");
 
-  if (!init) caml_memprof_init();
+  memprof_domain_t domain = Caml_state->memprof;
+  CAMLassert(domain);
+  CAMLassert(domain->current);
 
-  lambda = l;
-  if (l > 0) {
-    one_log1m_lambda = l == 1 ? 0 : 1/caml_log1p(-l);
-    rand_pos = RAND_BLOCK_SIZE;
-    /* next_rand_geom can be zero if the next word is to be sampled,
-       but rand_geom always returns a value >= 1. Subtract 1 to correct. */
-    next_rand_geom = rand_geom() - 1;
+  if (Sampling(thread_config(domain->current))) {
+    caml_failwith("Gc.Memprof.start: already started.");
   }
 
-  caml_memprof_renew_minor_sample();
-
-  callstack_size = sz;
-  started = 1;
+  /* Orphan any surviving tracking entries from a previous profile. */
+  if (!orphans_create(domain)) {
+    caml_raise_out_of_memory();
+  }
 
-  tracker = tracker_param;
-  caml_register_generational_global_root(&tracker);
+  double one_log1m_lambda = lambda == 1.0 ? 0.0 : 1.0/caml_log1p(-lambda);
+  /* If lambda is very small (but non-zero) then one_log1m_lambda can
+   * (for buggy implementations of caml_log1p) be positive and very
+   * large or infinite. This would cause chaos in the RNG, so we check
+   * against this and set one_log1m_lambda to negative infinity (which
+   * we can test for). We preserve the user's value of Lambda for
+   * inspection or debugging. */
+  if ((one_log1m_lambda > 0.0) || /* positive */
+      !(1.0 / one_log1m_lambda)) { /* one_log1m_lambda is infinite */
```

Comment:
```suggestion
  if (!(one_log1m_lambda <= 0.0)) {
```

This will catch all the ill-valued  `one_log1m_lambda`.
- If this is -oo, then nothing to do.
- If this is +oo or positive then comparison returns false.
- If NaN (who knows...), then comparison returns false.